scheduler/server/load_based_sched_alg.go

```diff
@@ -0,0 +1,780 @@
+
+	numTasksToStart int // number of tasks that can be started (when negative -> number of tasks to stop)
+	numWaitingTasks int // number of tasks still waiting to be started
+
+	tempEntitlement   int // temporary field to hold intermediate entitled num workers
+	tempNormalizedPct int // temporary field to hold the normalized load %
+
+}
+
+func (jc *jobClass) String() string {
+	return fmt.Sprintf("%s:TargetLoadPct:%d, origTasksWaiting:%d, origTasksRunning:%d, origTargetWorkers:%d, TasksToStart:%d, remainingWaitingTasks:%d, tempEntitlement:%d, tempNormalizedPct:%d",
+		jc.className,
+		jc.origTargetLoadPct,
+		jc.origNumWaitingTasks,
+		jc.origNumRunningTasks,
+		jc.origNumTargetedWorkers,
+		jc.numTasksToStart,
+		jc.numWaitingTasks,
+		jc.tempEntitlement,
+		jc.tempNormalizedPct)
+}
+
+// NewJobClass a job class with its target % worker load
+func NewJobClass(name string, targetLoadPct int) *jobClass {
+	return &jobClass{className: name, origTargetLoadPct: targetLoadPct, jobsByNumRunningTasks: map[int][]jobWaitingTaskIds{}}
+}
+
+// GetTasksToBeAssigned - the entry point to the load based scheduling algorithm
+// It returns the list of tasks that should be assigned to nodes as per the worker-class %s:
+// Allocate available workers to classes based on target % allocations for each class and the current and
+// running load for each class.
+// - When a class has tasks waiting to start, the algorithm will determing the number of workers the class it 'entitled' to:
+// the number of workers as per the class's target load %
+// - When classes are under-utilizing their 'entitlement', (due to lack of waiting tasks), the unallocated workers will be
+// ‘loaned’/used to run tasks from other classes (class allocations may exceed the targeted allocations %’s)
+// -The algorithm will try to allocate 100% of the workers (no unallocated reserves)
+//
+// When starting tasks within a class:
+// Jobs within a class are binned by the number of running tasks (ranking jobs by the number of active tasks).
+// When starting tasks for a class, the tasks are first pulled from jobs with the least number of active tasks.
+func (lbs *LoadBasedAlg) GetTasksToBeAssigned(jobsNotUsed []*jobState, stat stats.StatsReceiver, cs *clusterState,
+	jobsByRequestor map[string][]*jobState) ([]*taskState, []*taskState) {
+	log.Debugf("in LoadBasedAlg.GetTasksToBeAssigned: numWorkers:%d, numIdleWorkers:%d", len(cs.nodes), cs.numFree())
+
+	// make local copies of the load pct structures
+	lbs.classLoadPercents = lbs.LocalCopyClassLoadPercents()
+	lbs.requestorReToClassMap = lbs.getRequestorToClassMap()
+	lbs.classByDescLoadPct = lbs.getClassByDescLoadPct()
+
+	numWorkers := len(cs.nodes)
+	lbs.initOrigNumTargetedWorkers(numWorkers)
+
+	lbs.initJobClassesMap(jobsByRequestor)
+
+	rebalanced := false
+	var stopTasks []*taskState
+	if lbs.getRebalanceMinDuration() > 0 && lbs.getRebalanceThreshold() > 0 {
```

Comment:
Why not call getCurrentPercentsSpread here? If it's a perf concern, could you please comment that here.